Moved fastSim reconstruction to use era

Dr15Jones opened the pull request:
When trying to convert the reconstruction configurations from Sequences to Tasks, we found that the fastsim was greatly complicating the transition. One reason was the fastsim was directly rewriting Configuration.StandardSequences.Reconstruction_cff from within FastSimulation.Configuration.Reconstruction_AftMix_cff which was leading to very 'brittle' configuration code. For example, if someone accidently loaded Reconstruction_cff into the Process before Reconstruction_AftMix_cff the configuration would no longer work. Switching to fully using the fastSim era avoids many problems and makes fastsim behave like the other configuration modification workflows.

Dr15Jones committed:
Moved fastSim reconstruction to use era

cmsbuild commented:
The code-checks are being triggered in jenkins.

Dr15Jones commented:
I used release validation workflow 25402.0 as the test bed. I used the following simple configuration for testing
```
from TTbar_13TeV_TuneCUETP8M1_cfi_GEN_SIM_RECOBEFMIX_DIGI_L1_DIGI2RAW_L1Reco_RECO_EI_VALIDATION_DQM_PU import process
process.prune()
print process.dumpPython()
```
With that configuration, I found the my change returns a configuration with exactly the same module configuration as before. In addition, I checked that the Paths `reconstruction_step` and `reconstruction_befmix_step` contain the exact same list of modules.

Dr15Jones commented:
@gartung here is the fastsim change

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20884/1361

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

Configuration/StandardSequences
FastSimulation/Configuration
FastSimulation/Tracking
RecoEcal/EgammaClusterProducers
RecoEgamma/Configuration
RecoEgamma/EgammaPhotonProducers
RecoLocalCalo/Configuration
RecoLocalCalo/EcalRecProducers
RecoMET/Configuration
RecoMuon/Configuration
RecoMuon/GlobalMuonProducer
RecoMuon/GlobalTrackingTools
RecoMuon/MuonIdentification
RecoMuon/TrackingTools
RecoParticleFlow/PFTracking
RecoTracker/Configuration
RecoTracker/FinalTrackSelectors
RecoTracker/IterativeTracking
RecoTracker/MeasurementDet
TrackingTools/GsfTracking
TrackingTools/TrackRefitter


@perrotta, @civanch, @lveldere, @ssekmen, @mdhildreth, @cmsbuild, @franzoni, @slava77, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @TaiSakuma, @gouskos, @felicepantaleo, @jainshilpi, @abbiendi, @argiro, @Martin-Grunewald, @ahinzmann, @lgray, @varuns23, @seemasharmafnal, @mmarionncern, @battibass, @makortel, @jhgoh, @dgulhan, @jdolen, @HuguesBrun, @trocino, @rociovilar, @Sam-Harper, @GiacomoSguazzoni, @rovere, @VinInn, @jdamgov, @bellan, @nhanvtran, @gkasieczka, @schoef, @mschrode, @ebrondol, @amagitte, @echapon, @calderona, @cbernet, @gpetruc, @matt-komm, @mariadalfonso, @bachtis this is something you requested to watch as well.
@davidlange6, @slava77 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 50c957cdff49aec94e5b925f9f7aea7e71ac665f


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/dd401dfe30a0c9216a9ffb31b74db0c3c6e9416a
https://github.com/cms-sw/cmssw/commit/a2aeaf0f9952e731f4a712a24352a82b36755115
https://github.com/cms-sw/cmssw/commit/d9807b8c02751954924a75498347d587737ee879
https://github.com/cms-sw/cmssw/commit/da408a34bc951db5fa7bccb9695f1150fd69f795
https://github.com/cms-sw/cmssw/commit/e802724113c80a76b2b06eb1e8521d01a9773c68
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23655/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23655/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23655/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
1000.0 step3
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step3_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/dd401dfe30a0c9216a9ffb31b74db0c3c6e9416a
https://github.com/cms-sw/cmssw/commit/a2aeaf0f9952e731f4a712a24352a82b36755115
https://github.com/cms-sw/cmssw/commit/d9807b8c02751954924a75498347d587737ee879
https://github.com/cms-sw/cmssw/commit/da408a34bc951db5fa7bccb9695f1150fd69f795
https://github.com/cms-sw/cmssw/commit/e802724113c80a76b2b06eb1e8521d01a9773c68
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23655/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23655/git-merge-result

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

Dr15Jones commented:
The workflow 1000.0 is failing in the IBs with the exact same configuration error.

slava77 commented:
Changes in this PR should be discussed in some joint meeting with fastsim category managers.
This PR is effectively fusing two categories at the package assignment level and may have to require a joint signature or some other non-obvious solution.

ssekmen commented:
We have indeed planned to discuss the PR in the next simulation meeting when all simulation conveners are available.

makortel commented:
This PR conflicts with #20879, right?

Is FastSim planning to support both 2016 and 2017 detectors? (I already foresee the "fun" in maintaining tracking configurations for {2016, 2017} x {normal, FastSim}; ok, likely that would be "fun" within any system)

makortel commented on RecoTracker/IterativeTracking/python/MixedTripletStep_cff.py:
```diff
@@ -161,6 +169,18 @@
+#fastsim
+from FastSimulation.Tracking.SeedingMigration import _hitSetProducerToFactoryPSet
+_fastSim_mixedTripletStepSeedsA.seedFinderSelector.pixelTripletGeneratorFactory = _hitSetProducerToFactoryPSet(mixedTripletStepHitTripletsA)
+fastSim.toReplaceWith(mixedTripletStepSeedsA,_fastSim_mixedTripletStepSeedsA)
```
Ideally I'd like that these customizations would be placed after the other definitions of `mixedTripletStepSeedsA`. I presume the fact that `mixedTripletStepSeedsB` is cloned from `mixedTripletStepSeedsA` with a parameter not in FastSim configuration motivated the choice here.

How about something like this (following my suggestion of a similar ordering/parameter inconsistency with the XeXe tracking in #20749):
```python
_mixedTripletStepSeedsCommon = _seedCreatorFromRegionConsecutiveHitsTripletOnlyEDProducer.clone(...)
trackingLowPU.toModify(_mixedTripletStepSeedsCommon, ...)
mixedTripletStepSeedsA = _mixedTripletStepSeedsCommon.clone()
...
fastSim.toReplaceWith(mixedTripletStepSeedsA, _fastSim_mixedTripletStepSeedsA)

mixedTripletStepSeedsB = _mixedTripletStepSeedsCommon.clone(seedingHitSets = "mixedTripletStepHitTripletsB")
...
fastSim.toReplaceWith(mixedTripletStepSeedsB, _fastSim_mixedTripletStepSeedsB)
```

gartung replied:
I merged Chris' branch into my branch. I am testing my changes before I commit them on  my branch.

Dr15Jones replied:
@makortel This pull request came out of what we discovered in  #20879. I've kept the two seperate for now since they both can be a bit controversial. Plus, it is easier to test them independently to be sure they both do what they are intended.

makortel replied:
Ok, understood. But what about my comment for this piece of configuration?

slava77 commented on FastSimulation/Configuration/python/Reconstruction_AftMix_cff.py:
```diff
@@ -6,209 +6,6 @@
-# the actual import
-############################################
+#All work is now done with the fastSim era
 from Configuration.StandardSequences.Reconstruction_cff import *
```
here and in a few more places, is there a reverse equivalent, along the following lines:
 FastSimulation/Configuration will contain only the fastsim modifiers and by itself can not be used.
Instead then the standard include should change to Configuration.StandardSequences.Reconstruction_cff

Dr15Jones replied:
I was considering changing to just the standard includes as a seperate pull request. I wanted to minimize impact on other code (such as cmsDriver) as possible.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 50c957cdff49aec94e5b925f9f7aea7e71ac665f


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/e7a2ae1498eea390434bfaed6f32fcea53cf2855
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23688/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23688/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23688/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
1000.0 step3
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step3_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/e7a2ae1498eea390434bfaed6f32fcea53cf2855
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23688/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23688/git-merge-result

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

perrotta commented:
please test
(the latest IB, CMSSW_9_4_X_2017-10-11-2300, should finally contain the fix for the issue in the RelVal wf 1000)

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 50c957cdff49aec94e5b925f9f7aea7e71ac665f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23705/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/70922db3413f0934414865c37e735a97add7daef
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23705/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23705/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23705/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 7038 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2767051
 - DQMHistoTests: Total failures: 22319
 - DQMHistoTests: Total nulls: 297
 - DQMHistoTests: Total successes: 2744264
 - DQMHistoTests: Total skipped: 171
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 10 edm output root files, 26 DQM output files

slava77 commented:
@cmsbuild please test

in case the differences in comparisons go away

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones committed:
Keep fastSim changes to mixedTripletStepSeedsA near declaration

By using _mixedTripletStepSeedsACommon we can keep the fastSim changes
to mixedTripletStepSeedsA next to all era changes to the module. The
declaration of mixedTripletStepSeedsB is now relative to
_mixedTripletStepSeedsACommon.

cmsbuild commented:
The code-checks are being triggered in jenkins.

Dr15Jones commented:
@makortel  I believe I've implemented your suggested change.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #20884 was updated. @perrotta, @civanch, @lveldere, @ssekmen, @mdhildreth, @cmsbuild, @franzoni, @slava77, @davidlange6 can you please check and sign again.

makortel commented:
Thanks Chris.

Dr15Jones commented:
@makortel thanks you for the suggestion. I didn't like the fact that the fastSim era change was not near the original declaration.

cmsbuild commented:
-1

Tested at: 06a1da635e38312db6ed009a924c6881b0c622d7


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/561fe852773fa652d18c7cdc3c8845c0bd558045
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23732/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23732/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23732/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
10024.0 step5
<pre>runTheMatrix-results/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step5_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

</pre>10824.0 step5
<pre>runTheMatrix-results/10824.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+ALCAFull_2018+HARVESTFull_2018/step5_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+ALCAFull_2018+HARVESTFull_2018.log

</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/561fe852773fa652d18c7cdc3c8845c0bd558045
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23732/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23732/git-merge-result

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

perrotta commented:
please test
(usual unrelated crashes at step5)

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 06a1da635e38312db6ed009a924c6881b0c622d7
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23739/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/561fe852773fa652d18c7cdc3c8845c0bd558045
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23739/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23739/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23739/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 7038 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2767051
 - DQMHistoTests: Total failures: 22319
 - DQMHistoTests: Total nulls: 297
 - DQMHistoTests: Total successes: 2744264
 - DQMHistoTests: Total skipped: 171
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 10 edm output root files, 26 DQM output files

perrotta commented on RecoTracker/IterativeTracking/python/TobTecStep_cff.py:
```diff
@@ -233,16 +255,25 @@
     useHitsSplitting = True,
-    cleanTrajectoryAfterInOut = True
+    cleanTrajectoryAfterInOut = True,
+    TrajectoryCleaner = 'tobTecStepTrajectoryCleanerBySharedHits'
```
Also here, TrajectoryCleaner is moved inside the definition of tobTecStepTrackCandidates
Here it doesn't seems to touch Phase2 or anything else

perrotta reviewed (commented):
There are a lot of reco related modifications in both simulation and real data workflow (by the way, the only workflow which doesn't seem to be touched in the jenkins tests is the fastsim one, which is kind of wanted I'd say). This has to be inspected and fixed.

I went through the proposed configuration changes, looking for some obvious culprit, but I did not find any. The findings listed here are probably innocent (though I wonder what happens to Phase2 configs in RecoTracker/IterativeTracking/python/LowPtTripletStep_cff.py), but please keep also an eye on them.

Dr15Jones commented:
@perrotta is there a particular workflow you recommend I concentrate on to determine what caused the reco difference? I was going to dump the configurations with and without my changes to find any differences.

perrotta commented:
I see differencs in almost all workflows run by Jenkins.

Probably one in which I wouldn't really expect a FastSim change can  
have any effect is the RunSinglePh2017B wf 136.788:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_4_X_2017-10-12-1100+20884/23100/validateJR/all_OldVSNew_RunSinglePh2017B136p788/

Chris Jones <notifications@github.com> ha scritto:

> @perrotta is there a particular workflow you recommend I concentrate  
> on to determine what caused the reco difference? I was going to dump  
> the configurations with and without my changes to find any  
> differences.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/cms-sw/cmssw/pull/20884#issuecomment-336474082

Dr15Jones commented on RecoTracker/IterativeTracking/python/LowPtTripletStep_cff.py:
```diff
@@ -199,28 +210,38 @@
     doSeedingRegionRebuilding = True,
-    useHitsSplitting = True
+    useHitsSplitting = True,
+    TrajectoryCleaner = 'lowPtTripletStepTrajectoryCleanerBySharedHits'
```
I had originally moved it because it seemed innocuous. In fact, my change has no affect on `trackingPhase2PU140`. Before my change, `trackingPhase2PU140` would have modified `clustersToSkip` and `phase2clustersToSkip` but left `TrajectoryCleaner`. This era would still use the same python object instance for `lowPtTripletStepTrackCandidates` as the non-era version. Then after `trackingPhase2PU140` was changed, the call to 
```python
lowPtTripletStepTrackCandidates.TrajectoryCleaner = 'lowPtTripletStepTrajectoryCleanerBySharedHits'
```
would change `lowPtTripletStepTrackCandidates` both in the era and non-era case. Therefore moving setting of `TrajectoryCleaner` into the `clone` should make no functional difference.

Dr15Jones replied:
I also had to move setting of `lowPtTripletStepTrackCandidates.TrajectoryCleaner` to before the `fastSim` era change since the resulting `lowPtTripletStepTrackCandidates` object after the `fastSim` is applied is a module that does not have the parameter `TrajectoryCleaner`.

Dr15Jones committed:
Need to do deep copy of parameters

In order to avoid accidently changing the values from the original
module, we need to make deep copies of the parameters.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-20884/1452

cmsbuild commented:
Pull request #20884 was updated. @perrotta, @civanch, @lveldere, @ssekmen, @mdhildreth, @cmsbuild, @franzoni, @slava77, @davidlange6 can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
@perrotta I discovered that one of a fastsim functions was doing a shallow copy of a module's parameters. Then when a later piece of code changed the 'copy' it also changed the original. The most recent commit makes all configurations for workflow 10024.0 (which happens to be the one I started with) identical with and without this change.

Dr15Jones commented:
After the most recent change, I double checked workflow 25402.0 to make sure that I didn't break fastsim and it looks like it is still fine.

cmsbuild commented:
+1
Tested at: fd54ddfc630fb8498c3ac84d0bf02afd086081d6
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23763/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/25793ba9f63a57dc276e82bc720b8c8ed2f1eff8
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23763/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23763/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20884/23763/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2767139
 - DQMHistoTests: Total failures: 107
 - DQMHistoTests: Total nulls: 254
 - DQMHistoTests: Total successes: 2766607
 - DQMHistoTests: Total skipped: 171
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 10 edm output root files, 26 DQM output files

perrotta commented:
Please @Dr15Jones specify if this PR is needed by some other ongoing development, or if we can safely wait for its discussion at the next Sim meeting before finalizing the review and signing it

Dr15Jones commented:
It can wait for the next Sim meeting.

slava77 commented:
Looking at  FastSimulation/Configuration/python/Reconstruction_AftMix_cff.py in @20879,
I'd much rather this PR is integrated first.

The two PRs are clearly coupled.

makortel commented:
Below are some comments regarding changes in tracking configurations after some more thought.

We are not particularly happy of making the already-complex iterative tracking configuration files more complicated. Said that, this PR alone would be acceptable (even if barely) for us. But we have concerns what will happen when FastSim will start to support phase1 pixel+tracking (will 2016 support continue?) or phase2 at some point.

Were any alternative strategies thought (even if they would more or less break the era design "modify module where it is defined")? E.g. would something along the following work out in light of the Task migration (#20879)
```python
# In FastSimulation/Configuration/python/Reconstruction_cff.py

from Configuration.StandardSequences.Reconstruction_cff import *

<move all reco fastsim era customizations here>
```
? (basically following the current separation of FastSim and Reco but doing the reco customizations using the `fastsim` era)

slava77 commented:
On 10/25/17 9:58 AM, Matti Kortelainen wrote:
> Below are some comments regarding changes in tracking configurations
> after some more thought.
> 
> We are not particularly happy of making the already-complex iterative
> tracking configuration files more complicated. Said that, this PR alone
> would be acceptable (even if barely) for us. But we have concerns what
> will happen when FastSim will start to support phase1 pixel+tracking
> (will 2016 support continue?) or phase2 at some point.
> 
> Were any alternative strategies thought (even if they would more or less
> break the era design "modify module where it is defined")? E.g. would
> something along the following work out in light of the Task migration
> (#20879 <https://github.com/cms-sw/cmssw/pull/20879>)
> 
> # In FastSimulation/Configuration/python/Reconstruction_cff.py
> 
> from Configuration.StandardSequences.Reconstruction_cff import *
> 
> <move all reco fastsim era customizations here>
> 
> ? (basically following the current separation of FastSim and Reco but
> doing the reco customizations using the |fastsim| era)

Design of eras works best if the modifiers are applied in the same file
where the original modules are defined.

Perhaps it's time to think of refactoring the Iter_cff files?

Best would be to see the same configuration used for fastsim and fullsim
reco.
How far are we from that point?


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/20884#issuecomment-339247325>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbqHXpm0AiI9Pb7ren0woAPO1-o5Vks5svuo0gaJpZM4P0f5x>.
>

makortel commented:
> Perhaps it's time to think of refactoring the Iter_cff files?

Do you mean something like for each iterations splitting the seeding, pattern recognition+final fit, and selection to their own cff files imported in the main iteration cff file that also defines the iteration Task(/Sequence)?

slava77 commented:
On 10/25/17 10:58 AM, Matti Kortelainen wrote:
>     Perhaps it's time to think of refactoring the Iter_cff files?
> 
> Do you mean something like for each iterations splitting the seeding,
> pattern recognition+final fit, and selection to their own cff files
> imported in the main iteration cff file that also defines the iteration
> Task(/Sequence)?

I haven't thought of a specific change pattern.
It was more of a general statement


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/20884#issuecomment-339262971>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbgHwjhwDJ4bgWW9UhTruM8A5vSUnks5svvhBgaJpZM4P0f5x>.
>

Dr15Jones commented:
I agree with Slava
> Design of eras works best if the modifiers are applied in the same file where the original modules are defined.

and I will go further. Not putting the modifier in the original module can lead to unforseen side-effects. E.g. clones that do not match or sequences that break.

perrotta commented:
I have just verified that this PR interferes with #20666: @ssekmen please let us know which are the plans for the integration of the two PRs, priorities, etc.

ssekmen commented:
@perrota, discussing with the FastSim experts. We will come with a more concrete plan at our presentation at the tracking meeting on Nov 6.

perrotta commented:
Thank you Sezen.

At the same time, I have to correct myself, and apologize: #20884 and  
#20666 do NOT interfere each other, in the sense that I was able to  
merge them one on top of the other, without conflicts, starting from  
940pre3. I only badly interpreted the git output at a first glance.

Anyhow, having a concrete plan is good: let us know after nov 6. (At  
this point, I must imagine that they are both only meant to 10_0_X and  
the 2018+ releases.

Cheers,
Andrea

Sezen Sekmen <notifications@github.com> ha scritto:

> @perrota, discussing with the FastSim experts. We will come with a  
> more concrete plan at our presentation at the tracking meeting on  
> Nov 6.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/cms-sw/cmssw/pull/20884#issuecomment-340491188

perrotta commented:
+1
- Reco part of this PR was reviewed and approved
- Effect of this PR will likely be that the maintenance of the fastsim modules goes towards DPG/POGs instead of being kept separated in the fastsim package
- Several modules/sequences are marked with comments like "not commissioned and not relevant in FastSim": during the fastsim review it can be worth going through all those components and simply remove whatever not needed, thus simplifying and improving the future maintenance

davidlange6 commented:
hi @civanch - this is the PR discussed in the ORP yesterday

ssekmen commented:
+1

davidlange6 commented:
merge

kpedro88 commented:
@Dr15Jones I noticed some failures in fastsim premixing workflows in the IB tests (https://cms-sw.github.io/relvalLogDetail.html#slc7_aarch64_gcc700;CMSSW_10_0_X_2017-11-10-1100):
```
----- Begin Fatal Exception 10-Nov-2017 14:52:24 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
Exception Message:
EDAlias does not match data
There are no products of type 'recoTracks'
with module label 'mix' and instance name 'generalTracks'.
----- End Fatal Exception -------------------------------------------------
```

Might be caused by this PR, can you take a look?

kpedro88 commented:
ah, I see you were faster than me with #21263.
